_drafts/making-backend-scalable.markdown

```diff
@@ -0,0 +1,130 @@
+
+### Choice of Redis ###
+
+We started looking into possible solutions: traditional databases like MySQL, relational databases such as MongoDB and messaging queues such as RabbitMQ, Redis. Looking at the amount of data and the level of nestedness, we decided that we don't need a database with full-range of functionality and robustness. We needed something fast and simple to use. The implementation of our job queue is in the form of a Python dictionary, therefore, a fast key-value store like _Redis_ seemed like the right fit. Setting it up and playing around with it was a breeze, therefore we decided to go with this option.
+
+We are planning to add modules to support the other options in the future. More details about this will be give in the next sub-section.
+
+### Implementation ###
+
+While making the switch, we wanted to keep the ability to run Tango without dependency on the external service in order to keep the backwards competibility and ease of development environment setup.
+
+In order to achieve this we created (pseudo)abstract classes that include the common methods and decide on what is used under the hood based on the configuration.
+
+Since Python doesn't really have abstract classes, we have define methods that would initiate the appropriate class and return it.
+
+```
+# This is an abstract class that decides on 
+# if we should initiate a TangoRemoteDictionary or TangoNativeDictionary
+# Since there are no abstract classes in Python, we use a simple method
+def TangoDictionary(object_name):
+    if Config.USE_REDIS:
+        return TangoRemoteDictionary(object_name)
+    else:
+        return TangoNativeDictionary()
+```
+[See the full code here.](https://github.com/autolab/Tango/blob/master/tangoObjects.py#L232)
+
+In the future, we can have other implementations, such as MongoDB, and just add it as an option here. For example: 
+
+```
+def TangoDictionary(object_name):
+    if Config.USE_REDIS:
+        return TangoRemoteDictionary(object_name)
+    elif Config.USE_MONGO:
+        return TangoMongoDictionary(object_name)
+    else:
+        return TangoNativeDictionary()
+```
+
+
+
+### Serialization and Marshalling ###
+
+Redis, being a key-value store, only accepts strings as values. Even though the objects we are trying to store are quite simple and flat, we do not want to create a new key-value pair for every attribute of the class. Instead we decided to serialize what we call the 'remote objects' using the `pickle` module included in the standard python library.
+
+A sample method to put the object into the store looks like
+
+```
+def set(self, id, obj):
+    pickled_obj = pickle.dumps(obj)
+    self.r.hset(self.hash_name, str(id), pickled_obj)
+    return str(id)   
+```
+
+An to get it back you would need:
```

Comment:
- "And" is misspelled, but I think it reads better if you drop it altogether:

- __To get it back__...